File: java/src/test/java/dev/drim/relay/harness/LlmHarness.java
package dev.drim.relay.harness;

import dev.drim.relay.domain.Events;
import dev.drim.relay.seams.SummaryModel;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * The canonical FAKE (04-dependencies.md §6): the LLM is nondeterministic, paid, and external, so
 * the boundary is a deliberate in-process double, not a container. Seed = program the next response
 * (canned / empty / oversized); Assert = interaction verification — the captured request is where
 * the prompt-injection catch lives; Reset = clear responses + captured calls. Hand-rolled on
 * purpose (no mocking framework) so the pattern reads cross-language. Mirrors go/harness/llm.go.
 */
public final class LlmHarness implements DependencyHarness {
  private final Object lock = new Object();
  private final Deque<String> programmed = new ArrayDeque<>();
  private final List<Events.SummaryRequest> captured = new ArrayList<>();

  @Override
  public void start() {}

  @Override
  public void reset() {
    clear();
  }

  @Override
  public void stop() {}

  /** Returns the fake as the app's {@link SummaryModel} seam. */
  public SummaryModel model() {
    return new FakeSummaryModel();
  }

  /** Seeds the next response (FIFO). Unprogrammed → a canned summary. */
  public void programResponse(String response) {
    synchronized (lock) {
      programmed.addLast(response);
    }
  }

  /** Returns the requests the app made — for interaction verification. */
  public List<Events.SummaryRequest> capturedRequests() {
    synchronized (lock) {
      return new ArrayList<>(captured);
    }
  }

  public void clear() {
    synchronized (lock) {
      programmed.clear();
      captured.clear();
    }
  }

  private final class FakeSummaryModel implements SummaryModel {
    @Override
    public String complete(Events.SummaryRequest request) {
      synchronized (lock) {
        captured.add(request);
        if (!programmed.isEmpty()) {
          return programmed.removeFirst();
        }
        return "(canned summary)";
      }
    }
  }
}

File: dotnet/harness/Relay.Testing/LlmHarness.cs
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Mvc.Testing;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Relay.Api.Features.Assistant;

namespace Relay.Testing;

/// <summary>
/// LLM harness — the canonical FAKE (spec/04-dependencies.md §6): nondeterministic, paid
/// and external, so the boundary is a deliberate in-process double, not a container.
/// <c>Seed</c> = program the next response (canned / empty / oversized);
/// <c>Assert</c> = interaction verification — the captured request is where the prompt-
/// injection catch lives; <c>Reset</c> = clear programmed responses and captured calls.
/// Hand-rolled on purpose (no mocking framework) so the pattern reads cross-language.
/// </summary>
public sealed class LlmHarness<TProgram> : IDependencyHarness<TProgram>
    where TProgram : class
{
    private readonly FakeSummaryModel _fake = new();

    public void ConfigureWebHostBuilder(IWebHostBuilder builder) =>
        builder.ConfigureServices(services =>
        {
            services.RemoveAll<ISummaryModel>();
            services.AddSingleton<ISummaryModel>(_fake);
        });

    public Task Start(WebApplicationFactory<TProgram> factory, CancellationToken cancellationToken) =>
        Task.CompletedTask;

    public Task Stop(CancellationToken cancellationToken) => Task.CompletedTask;

    public void ProgramResponse(string response) => _fake.ProgramResponse(response);

    public IReadOnlyList<SummaryModelRequest> CapturedRequests => _fake.CapturedRequests;

    public void Reset() => _fake.Clear();
}

public sealed class FakeSummaryModel : ISummaryModel
{
    private readonly Lock _lock = new();
    private readonly List<SummaryModelRequest> _captured = [];
    private readonly Queue<string> _programmed = new();

    public Task<string> Complete(SummaryModelRequest request, CancellationToken ct)
    {
        lock (_lock)
        {
            _captured.Add(request);
            return Task.FromResult(_programmed.TryDequeue(out var response) ? response : "(canned summary)");
        }
    }

    public IReadOnlyList<SummaryModelRequest> CapturedRequests
    {
        get
        {
            lock (_lock)
            {
                return [.. _captured];
            }
        }
    }

    public void ProgramResponse(string response)
    {
        lock (_lock)
        {
            _programmed.Enqueue(response);
        }
    }

    public void Clear()
    {
        lock (_lock)
        {
            _captured.Clear();
            _programmed.Clear();
        }
    }
}
